### ISSUE TITLE
Unclickable links on page

### ISSUE THREAD

**Issue by technion** [servo/servo#16405]
Hi,

Great work on the Windows release!

I downloaded the nightly Windows build today, unfortunately the website I cannot get working is my own: https://lolware.net

None of the links on the page are clickable (Except the GitHub link in the top right, which is fine for some reason).

**Comment by jdm** [comment]
The links are clickable if the page is not scrolled. Once the page is scrolled away from the root view, they stop being clickable.

**Comment by jdm** [comment]
I see this on other pages too - any scroll offset in the page does not affect hit testing in the way that's expected. This might be the same as #16189.

**Comment by jdm** [comment]
This is observable in the following testcase:
```html
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
```

**Comment by jdm** [comment]
@mrobinson Is this something you could look into?

**Comment by cjthompson** [comment]
This issue exists in the macOS build also, so it's not unique to the Windows build.  An easy test is Hacker News.  Just scroll the page and the mouse position still things it's relative to the top of the page instead of the current view.

**Comment by jdm** [comment]
We can even make an automated testcase out of the previous one that I constructed:
```html
<a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<p><a href="http://google.com">this is a link</a>
<br><br>
<script>
  var firstLinkRect = document.querySelector("a").getBoundingClientRect();
  console.log(document.elementFromPoint(firstLinkRect.x, firstLinkRect.y).tagName);
  window.scrollBy(0, 1);
  console.log(document.elementFromPoint(firstLinkRect.x, firstLinkRect.y - 1).tagName);
</script>
```
On master this shows:
```
A
HTML
```

**Comment by mrobinson** [comment]
Currently the way that Servo it written it seems that the scroll position requesting from script on the root is handled entirely within Servo and not in WebRender. This is likely because scrolling in WebRender is handled asynchronously. This means that although Servo does ask WebRender to scroll the root node, the offset of the root is tracked and applied entirely within Servo.

The most pressing problem here in this test case is that the scroll offset that Servo is using internally is completely bogus. Since Servo doesn't actually do real scrolling itself, it isn't doing any real calculation about what the request means in reality (taking into account scroll node height and window height, etc). In this test case the offset that Servo ends up requesting is (-16,1). The 1 in this case is expected, but the -16 comes from this bit of code in window.rs:
```rust
    pub fn scroll(&self, x_: f64, y_: f64, behavior: ScrollBehavior) {
        // Step 3
        let xfinite = if x_.is_finite() { x_ } else { 0.0f64 };
        let yfinite = if y_.is_finite() { y_ } else { 0.0f64 };

        // Step 4
        if self.window_size.get().is_none() {
            return;
        }

        // Step 5
        //TODO remove scrollbar width
        let width = self.InnerWidth() as f64;
        // Step 6
        //TODO remove scrollbar height
        let height = self.InnerHeight() as f64;

        // Step 7 & 8
        //TODO use overflow direction
        let body = self.Document().GetBody();
        let (x, y) = match body {
            Some(e) => {
                let content_size = e.upcast::<Node>().bounding_content_box_or_zero();
                let content_height = content_size.size.height.to_f64_px();
                let content_width = content_size.size.width.to_f64_px();

                (xfinite.max(0.0f64).min(content_width - width),
                 yfinite.max(0.0f64).min(content_height - height))
            },
            None => {
                (xfinite.max(0.0f64), yfinite.max(0.0f64))
            }
        };
```

`inner_width` is 1024, but `content_width` is 1008 (which is the width of the first child flow). I get the impression that the line:
```rust
                (xfinite.max(0.0f64).min(content_width - width),
```
should be
```rust
                (xfinite.min(content_width - width).max(0.0f64),
```
but I cannot say for sure. It's a bit odd that the content_width is coming from the first child flow instead of the root flow, but perhaps this expected. In any case, the end result is that when doing the hit test, the -16 is applied to the X coordinate of the point. This code hasn't changed recently, so perhaps something in layout changed recently. So there are one or maybe two bugs here.

In addition to this problem, it seems that when sending scroll offsets for the root scroll node to WebRender, Servo isn't properly forming the ClipId for the root node. This isn't causing the test to fail and probably isn't the source of the original issue (since it only matters with scrolls from script), but does mean that WebRender isn't scrolling the root node properly. This was an oversight on my part and I'm going to make a change in WebRender to try to fix this problem and prevent ones like it in the future.

## Asynchronous Scrolling

I think it's worth writing a small aside about the issues of asynchronous scrolling in WebRender here. When the script in the test case sends the scroll offset to WebRender, it hasn't even constructed the ClipScrollTree for the first frame. That means that even if Servo was requesting the new scroll offsets right away (which it isn't: probably yet another bug), they would be empty. This is because the offsets don't include any pending scroll offsets (probably another bug). The issue with fixing it is illustrated by the problem in Servo though. WebRender cannot know what the "real" scroll offsets are going to be until it has the ClipScrollTree and figures out the final scroll offsets taking into account node heights and window sizes and all that. Maybe it doesn't make sense to parrot back any old bogus offset that it receives from Servo. It's worth thinking about.

So in summary, there are maybe four or five bugs here, but only one or two that seem ultimately important for @jdm's test case. On the other hand, the bugs and the test case may not relate to the original problem (which I haven't been able to reproduce yet). I'll investigate that a bit more and report back.

Paging @emilio simply because he might be able to answer my question about the Flow node width. Sorry for the lengthy comment. This was a very complicated situation.

**Comment by mrobinson** [comment]
I created https://github.com/servo/servo/pull/16529 to fix one issue that I mentioned.

**Comment by glennw** [comment]
@mrobinson @emilio It seems that the issue with clicking links once scrolled is a (relatively) recent regression over what we had previously.

Is it possible to fix that regression without doing all the "proper" fixes that @mrobinson has mentioned above? Or is this a case where the recent WR scrolling changes have exposed these issues and they need to be fixed properly, perhaps?

**Comment by jonathandturner** [comment]
Did a `git bisect`.  Looks like the breakage happened at or near fe75382bcbb20efe43e650d144fa3bde7160f853

**Comment by jonathandturner** [comment]
@glennw / @mrobinson ^

**Comment by jonathandturner** [comment]
The other two checkouts in the bisect before I hit issues: 2f6cf2e9c883d919b35d952f47d6bb1670f21def and 9d72e89ce3502cb58147375c8ecc4e9e9356b1ee

**Comment by mrobinson** [comment]
I have created https://github.com/servo/servo/pull/16807 to fix the issue uncovered by @jdm's test. I have an idea what might be causing the issue with clickable links and will investigate a bit to see if my hypothesis is correct.

**Comment by jonathandturner** [comment]
Unfortunately, #16807 doesn't seem to fix the issue. You can see the issue by opening http://news.ycombinator.com and scrolling down the page. Anywhere you click, it's like you're still on the first page.

**Comment by mrobinson** [comment]
I think I have a fix for this and hope to have it ready to post in the next couple days.

**Comment by jonathandturner** [comment]
@mrobinson - great! Would happy to test when you have something ready

**Comment by jonathandturner** [comment]
@mrobinson - Btw, I'd also be happy to help fix it since it's blocking me from some of the tests I'm trying to do, if you could spare a few minutes to point me in the right direction.

**Comment by mrobinson** [comment]
@jonathandturner: I've posted a PR now. If you have time to test it, that would be awesome!

**Comment by sviter** [comment]
This issue is still present in latest servo windows nightly build

**Comment by mrobinson** [comment]
@sviter Is there a particular page where you see this issue?

**Comment by sviter** [comment]
@mrobinson I've tried wikipedia, servo and rust web-sites from start page

**Comment by cbrewster** [comment]
The windows nightly upload is currently broken, so you are probably using an out of date version. See #16560

**Comment by jonathandturner** [comment]
Unfortunately, yes, windows nightly isn't currently updating. Hoping we can get that fixed soon. To see this fix, you currently have to build from source.
